adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java

```diff
@@ -0,0 +1,827 @@
+            throws AuthenticationException {
+        final URL authorityUrl = StringExtensions.getUrl(authenticationRequest.getAuthority());
+        if (authorityUrl == null) {
+            throw new AuthenticationException(
+                    ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_URL);
+        }
+
+        // validate authority
+        if (mAuthContext.getValidateAuthority()) {
+            validateAuthority(authorityUrl);
+            if (!mAuthContext.getIsAuthorityValidated()) {
+                Logger.v(TAG, "Invalid authority, sending back exception via callback.");
+                throw new AuthenticationException(
+                        ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_INSTANCE);
+            }
+        }
+
+        // Verify broker redirect uri for non-silent request
+        if (canSwitchToBroker(authenticationRequest) && !authenticationRequest.isSilent()) {
+            // Verify redirect uri match what we expecting for broker non-silent request. For silent request,
+            // since we don't require developer to pass us the redirect uri, we won't perform the validation.
+            if (!authenticationRequest.isSilent()) {
+                verifyBrokerRedirectUri(authenticationRequest);
+            }
+        }
+    }
+
+    /**
+     * Perform authority validation.
+     * True if the passed in authority is valid, false otherwise.
+     */
+    private void validateAuthority(final URL authorityUrl) {
+        if (!mAuthContext.getIsAuthorityValidated()) {
+            Logger.v(TAG, "Start validating authority");
+            mDiscovery.setCorrelationId(mAuthContext.getRequestCorrelationId());
+            final boolean authorityValidationResult = mDiscovery.isValidAuthority(authorityUrl);
+            Logger.v(TAG, "Finish validating authority:" + authorityUrl + " result:" + authorityValidationResult);
+            mAuthContext.setIsAuthorityValidated(mDiscovery.isValidAuthority(authorityUrl));
+        }
+    }
+
+    /**
+     * 1. For Silent flow, we should always try to look local cache first.
+     *    i> If valid AT is returned from cache, use it.
+     *    ii> If no valid AT is returned, but RT is returned, use the RT.
+     *    iii> If RT request fails, and if we can talk to broker, go to broker and check if there is a valid token.
+     * 2. For Non-Silent flow.
+     *    i> Do silent cache lookup first, same as 1.
+     *       a) If we can talk to broker, go to broker for auth.
+     *       b) If not, launch webview with embedded flow.
+     *
+     * If silent request succeeds, we'll return the token back via callback.
+     * If silent request fails and no prompt is allowed, we'll return the exception back via callback.
+     * If silent request fails and prompt is allowed, we'll prompt the user and launch webview.
+     */
+    private void acquireTokenAfterValidation(final CallbackHandler callbackHandle,
```

Comment:
refactor this method into tryAcquireTokenSilent and tryAcquireTokenInteractively, the detailed logic determining should I try silent or interactive will go inside each method.